api/src/services/users.js

```diff
@@ -577,71 +652,24 @@ module.exports = {
       }
     }
 
-    return Promise
-      .all([
-        getUpdatedUserDoc(username, data),
-        getUpdatedSettingsDoc(username, data),
-      ])
-      .then(([ user, settings ]) => {
-        const tokenLoginError = tokenLogin.validateTokenLogin(data, false, user, settings);
-        if (tokenLoginError) {
-          return Promise.reject(error400(tokenLoginError.msg, tokenLoginError.key));
-        }
+    const [user, userSettings] = await Promise.all([
+      getUpdatedUserDoc(username, data),
+      getUpdatedSettingsDoc(username, data),
+    ]);
 
-        const response = {};
-
-        return Promise.resolve()
-          .then(() => {
-            if (data.place) {
-              settings.facility_id = user.facility_id;
-              return places.getPlace(user.facility_id);
-            }
-
-            if (_.isNull(data.place)) {
-              if (settings.roles && auth.isOffline(settings.roles)) {
-                return Promise.reject(error400(
-                  'Place field is required for offline users',
-                  'field is required',
-                  {'field': 'Place'}
-                ));
-              }
-              user.facility_id = null;
-              settings.facility_id = null;
-            }
-          })
-          .then(() => {
-            if (data.contact) {
-              return validateContact(settings.contact_id, user.facility_id);
-            }
-
-            if (_.isNull(data.contact)) {
-              if (settings.roles && auth.isOffline(settings.roles)) {
-                return Promise.reject(error400(
-                  'Contact field is required for offline users',
-                  'field is required',
-                  {'field': 'Contact'}
-                ));
-              }
-              settings.contact_id = null;
-            }
-          })
-          .then(() => db.users.put(user))
-          .then(resp => {
-            response.user = {
-              id: resp.id,
-              rev: resp.rev
-            };
-          })
-          .then(() => db.medic.put(settings))
-          .then(resp => {
-            response['user-settings'] = {
-              id: resp.id,
-              rev: resp.rev
-            };
-          })
-          .then(() => tokenLogin.manageTokenLogin(data, appUrl, response))
-          .then(() => response);
-      });
+    const tokenLoginError = tokenLogin.validateTokenLogin(data, false, user, userSettings);
+    if (tokenLoginError) {
+      return Promise.reject(error400(tokenLoginError.msg, tokenLoginError.key));
+    }
+
+    await validateUserFacility(data, user, userSettings);
+    await validateUserContact(data, user, userSettings);
+    const response = {
+      user: await saveUserUpdates(user),
+      'user-settings': await saveUserSettingsUpdates(userSettings),
+    };
+
+    return tokenLogin.manageTokenLogin(data, appUrl, response);
```

Comment:
I've flatten this function a bit by taking out some logic into new functions.
And I think the tokenLogin service is returning the same response object 🤔 this line should be okay

Reply:
Nice work on this refactor. This was one of the functions that failed the complexity rule, actually, so we that was needed anyway!